Second version of well completions

aprilzhang opened the pull request.

aprilzhang commented:
Still in progress of resolving conflicts.

aprilzhang commented:
ready for code review

alexandruandrei91 reviewed (approved):
Looks good

rubenthoms commented on src/lib/components/WellCompletions/utils/dataUtil.ts:
```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Util method to prepare stratigraphy and well data from the given time step range and other settings for plotting
+ * @param stratigraphy
+ * @param wells
+ * @param range
+ * @param rangeDisplayMode
+ * @param hideZeroCompletions
+ * @returns
```
Looks like an incomplete documentation? I do actually prefer the documentation style you used further up with the comments in the interface definition.

aprilzhang replied:
further up with the comments in the interface definition?

rubenthoms reviewed (changes requested):
Very nice work :) Like your coding style. Clean and very comprehensible. Made some comments/suggestions in the code. However, while testing, I found two inconveniences, both related to the tool tips:

1) When hovering an element located close to the right border of the window, the tooltip gets cut off. Suggestion: check if `(clientX + width) > windowWidth` and display the dialog on the left side of the cursor when true. Same accounts for the vertical direction.

![image](https://user-images.githubusercontent.com/69145689/112021530-ae5e3500-8b31-11eb-8a8f-6b83296c51f2.png)

2) When hovering the axis labels in x-direction, the tool tip's labels and values are separated by different font weights. That makes it easier to differentiate between them. When hovering the well bars, however, there is no such separation, making it harder for the user to get the information at a glance. I suggest to also make use of the font weight here.

Furthermore, some suggestions regarding the UI:
1) The well tool tip labels start with a capital letter while the x axis' tool tip labels start with a small letter. I think it would be nice if they were capital both places.

2) The time range slider's current value text should somehow have a background and/or only be shown while dragging the slider. Now it looks weird when they are overlapping.
![image](https://user-images.githubusercontent.com/69145689/112022861-f762b900-8b32-11eb-9760-521f648f7b2b.png)

3) The `Well sorting levels` dialog is not that comprehensive to me on first sight. Maybe you could add some placeholder text to the table or make the add option more obvious?
![image](https://user-images.githubusercontent.com/69145689/112023176-490b4380-8b33-11eb-822b-d7ac19ea382d.png)

And finally, something that might be a bug. When hovering, sometimes, the tool tips are not shown (when zone > 14.2) and sometimes the zone in the tool tip is not corresponding to the one on the y axis (not sure how it should be, though) [which might be related to the first issue].

![image](https://user-images.githubusercontent.com/69145689/112027282-532f4100-8b37-11eb-8b22-4bfc56a31b6b.png)


Regarding the PR, I don't know if you already intend it, but afaik it's preferred to squash all commits into a single one with a detailed description. Not sure if this applies to this PR as well, @anders-kiaer?

aprilzhang commented:
Thank you @rubenthoms , your comments are very useful : )
I have made the following updates:
A.1: I changed to use react-tooltip which has native support for checking the edges.
A.2: Fixed.
B.1: Fixed.
B.2: This is a known issue with the slider from equinor design system. It is still useful to use it rather than material UI for its font and theme etc. I will report this to the design system repo.
B.3: Done.
I couldnt quite reproduce the tooltip bug in the screenshot, I did notice the tooltip wont show up if the menu is open though.
The squash and merge option is available when the PR is merged.
![image](https://user-images.githubusercontent.com/1658235/112454058-36ba2100-8d50-11eb-91fa-db622ae3c178.png)

aprilzhang commented on src/lib/components/WellCompletions/WellCompletions.jsx:
```diff
@@ -8,6 +8,7 @@ import PropTypes from "prop-types";
 import React from "react";
 import WellCompletionComponent from "./components/WellCompletionComponent";
 
+export const DataContext = React.createContext();
 const WellCompletions = props => {
     return <WellCompletionComponent id={props.id} data={props.data} />;
 };
```
Thanks!

aprilzhang commented on src/lib/components/WellCompletions/components/Common/useDialog.tsx:
```diff
@@ -0,0 +1,94 @@
+/* eslint-disable @typescript-eslint/no-empty-function */
+import { Dialog } from "@material-ui/core";
```
Done

rubenthoms commented:
Great work, looks nice :) 👍 However, I still get the bug with the tooltip not appearing/showing wrong values. I investigated a bit more and found that it only happens after resizing the window (maybe the coordinates are somehow off then?). 
I tested in Chrome and Firefox. While testing in the latter I also found the y-axis labels to be cut off (only in Firefox, not in Chrome).
![image](https://user-images.githubusercontent.com/69145689/112625108-c33b1100-8e2e-11eb-827c-cec21a591667.png)

rubenthoms reviewed (approved):
Works like a charm :) The only thing you could still consider is fixing the cutoff of the y-axis labels in Firefox, but not super important.

LGTM 👍
